x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformRetryableActions.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.transform.transforms;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.RetryableAction;
+import org.elasticsearch.common.util.CancellableThreads;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.transform.notifications.TransformAuditor;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * Maintains the state of all currently scheduled retry tasks for all Transforms on this node.
+ * This state is not expected to be persisted across the cluster or persisted across cluster restarts.
+ * This class is thread safe.
+ */
+public class TransformRetryableActions {
```

Comment:
Yeah, we could, I picked Map + `RetryableAction` instead of `TransformScheduler` mostly because I didn't want to modify `TransformScheduler` to have the new needed functionality:
1. The ability for a `TransformScheduledTask` to mark itself "finished" and remove itself from the queue, which we could do by having the triggered listener return a boolean to say if it should keep running.
2. The random interval between retries, which would be fairly easy to write since we always calculate the next execution time.
3. Some way for users to visibly see that a retry is happening while the persistent task is started but the transform indexer is not, which would be some new API in the scheduler.

RetryableAction already had the pattern for 1 and 2, so it just needed 3 and the global storage (Map).

Now that we have a working version, I could maybe timebox another iteration that uses `TransformScheduler` with the above changes.  Having fewer moving parts would be good.

Reply:
This was kinda my result of this: https://github.com/prwhelan/elasticsearch/commit/4ea9842e404d132e3ca73708ce63d83a9771a062

There's no 2 or 3 yet, but that's about what it would look like.  There may be some more adjustments within the Scheduler to not log/audit, otherwise every transform start would see audits for register(startup) -> deregister(startup) -> register(start)

Reply:
> Now that we have a working version, I could maybe timebox another iteration that uses TransformScheduler with the above changes. Having fewer moving parts would be good.

I agree. Since we already have `TransformScheduler` class (and, BTW, we own it so we can extend it with any functionality that makes sense, e.g.: add random jitter to scheduled time), it would be good to re-use it not to introduce yet another retrying mechanism.